### PR TITLE
test: make the -devsyntheticcoins group addresses distinct

### PR DESCRIPTION
**Pull request by PrestackI**
The `-devsyntheticcoins=<n>[:<groups>]` dev harness advertises `<n>` coins "spread over `<groups>` addresses" — `interfaces/wallet_coin_source.h:166-168`, and the arg help at `bitcoin.cpp:186-189`. It could not deliver that: each group address was derived as

```cpp
std::string("SynthGroup") + std::string(1, static_cast<char>('A' + (g - 1) % 26)) + "000...";
```

so every `g >= 27` aliased back onto `'A'`. `r.group_address = address` (`synthetic_coin_source.cpp:163`) and `buildGroups` keys `m_groups[r.group_address]` (`coinviews.cpp:176`), so the store saw **at most 27 groups** no matter what was requested.

Nothing clamps in between — `ParseInt32` (`bitcoin.cpp:1440-1445`) is unbounded and `seedSynthetic` moves the records in verbatim — so the shortfall was silent, and the log line at `:174-175` reported the requested count regardless. `-devsyntheticcoins=50000:20000` seeded a whale of 10002 plus 26 groups of ~1538 and logged "seeded 50000 coins over 20000 groups".

The consequence is that the many-groups axis the #3183 acceptance harness documents has never been exercisable. It surfaced while measuring #3228 item 2, which names exactly that shape as its first step.

## Change

Render the group index in full, zero-padded to the same 34-character width as the whale address, so the mapping is injective. `git grep SynthGroup` has one hit, so nothing depended on the letter shape.

Also corrects two stale statements that the old behaviour made wrong: the header contract, and the "the remaining groups get 1000 each" comment (`small_group` is `min(1000, total/groups)`, so it falls to an even split as `<groups>` grows, and to 1 apiece when they are equal).

## Test

New case in `qt_walletcoinstore_tests` pins the directory size and address injectivity at 30 and 500 groups — both across the old 27 ceiling — and checks the seeded coins are all still accounted for. Against the previous code it reports:

```
check static_cast<int>(dir.size()) == groups has failed [27 != 30]
check static_cast<int>(dir.size()) == groups has failed [27 != 500]
```

https://claude.ai/code/session_01Q5mqyoeyLygRtQTmUqPTTg

### CODE REVIEW

**Review by jamescowens — COMMENTED** [review]
Verified the diagnosis end to end. Every anchor is accurate, including the arithmetic.

| Claim | Confirmed |
|---|---|
| Arg advertises "<n> coins over <groups> address groups" | `qt/bitcoin.cpp:186-189` |
| `ParseInt32` is unbounded; nothing clamps between the arg and the seed | `qt/bitcoin.cpp:1440-1447` — throws on a malformed parse, no range check |
| `buildGroups` keys the directory on `group_address` | `wallet/coinviews.cpp:176` |
| The log reports the *requested* counts | `synthetic_coin_source.cpp:174-175`, `total_coins, groups` |
| `git grep SynthGroup` has one hit | one, the replaced line |

I reproduced the worked example exactly: `50000:20000` gives `small_group = min(1000, 2) = 2`, `whale = 50000 - 19999*2 = 10002`, and the 19999 non-whale groups fold onto 26 addresses at ~1538 coins each. That is the number in the description.

The fix is right and minimal. All three addresses stay 34 characters (`SynthWhale` + 24, `SynthGroup` + 24 digits), and 24 digits is unconditionally wider than any `int`, so `%024d` is injective for every value `groups` can hold — no truncation regime to worry about. The whale cannot collide either, different prefix.

## Finding: the same silent shortfall survives one regime over

The aliasing ceiling is gone, but the *failure mode* the description opens with — a harness that quietly delivers fewer groups than asked for, while logging the requested number — is still reachable. It just moves from `groups > 26` to `groups > total_coins`.

`small_group` is `min(1000, total_coins / groups)`, integer division. Once `groups` exceeds `total_coins` that is **0**, every non-whale group seeds zero records, and since `buildGroups` only ever sees records, those groups never reach the directory at all:

```
total=5000  groups=5000   small_group=1   whale=1      dir_rows=5000
total=5000  groups=6000   small_group=0   whale=5000   dir_rows=1     <-- shortfall
total=100   groups=500    small_group=0   whale=100    dir_rows=1     <-- shortfall
```

So `-devsyntheticcoins=100:500` builds a one-row directory and logs `seeded 100 coins over 500 groups`. That is the identical trap, and it costs the next person measuring a many-groups shape the same afternoon this PR was written to give back.

This also makes the newly strengthened header sentence inaccurate in exactly the regime it is about:

> the group addresses are distinct for every `<groups>`, so a large group count produces a correspondingly large directory

The first clause is now true. The second does not follow from it — injective addresses are necessary but not sufficient, because an address with no records never becomes a directory row. The `.cpp` comment stops at "and to 1 apiece when they are equal", which is the last point at which the claim still holds, so the boundary is right there and just not carried through.

The test does not pin it either: `total = 5000` against 30 and 500 groups leaves `small_group` at 166 and 10, both comfortably inside the safe regime. Nothing currently fails if `small_group` goes to zero.

Cheapest fix that kills the *silent* part rather than the shortfall — log what was actually built:

```cpp
LogPrintf("SyntheticCoinSource: seeded %d coins over %d groups (whale group: %d)",
          total_coins, <actual directory size>, whale);
```

Or clamp at the top, which preserves the contract instead of just reporting its violation:

```cpp
groups = std::clamp(groups, 1, total_coins);
```

with a log line when it bites. I would lean to the clamp, since it makes the header sentence true as written rather than requiring another caveat, and `groups > total_coins` has no sensible meaning anyway. Either way a case at `groups > total_coins` in the new test would pin the boundary the same way the 30/500 cases pin the old one.

Not a blocker — the PR strictly improves things and the regime is out of the way. But it is the same shape of bug, still live, in the file being fixed for it.

## Non-gating

`-devsyntheticcoins` is a dev-only GUI harness argument; the synthetic source substitutes for the wallet scan only. Nothing here is reachable from validation, the wallet, or a default build's runtime path.

ACK. Worth folding in the clamp or the honest log line while you are here.

**Review by jamescowens — APPROVED** [review]
Verified the injectivity test asserts all three properties that matter (directory size == groups, address set size == groups, coin conservation) across the old 27-group aliasing boundary, and the 5000-at-clamp input is a genuine off-by-one guard on the clamp bound. Doc and -devsyntheticcoins help text updated to match. Dev-only surface; ran in the integration tree suite.

File: src/wallet/synthetic_coin_source.cpp [hunk 1]
```diff
@@ -147,10 +149,16 @@ SyntheticCoinSourceImpl::SyntheticCoinSourceImpl(int total_coins, int groups)
...
+        // injective: an aliasing scheme (a single letter mod 26, say)
+        // collapses the directory onto 26 addresses no matter what
+        // <groups> asks for, because buildGroups() keys on group_address
+        // -- which silently caps the many-groups axis this harness exists
+        // to exercise.
         const std::string address = (g == 0)
             ? std::string("SynthWhale000000000000000000000000")
-            : std::string("SynthGroup") + std::string(1, static_cast<char>('A' + (g - 1) % 26))
-                  + "00000000000000000000000";
+            : strprintf("SynthGroup%024d", g);
```
[the remark this answers is not in the document]
Comment (jamescowens) [reply src/wallet/synthetic_coin_source.cpp:177]: Addressed by 5dfd60a83 (pushed a few hours after this comment, before merge): <groups> is now clamped to <total_coins> up front, which makes small_group = min(1000, total/groups) >= 1 post-clamp, so every group seeds at least one record and the directory holds exactly the clamped group count -- your 10:20 example yields 10 groups with the clamp logged ("clamped %d requested groups to %d"). The help text and the header doc were updated in the same commit, and the test pins the clamp at its boundary (groups == total == 5000, which fails against a mutated total_coins - 1 bound). One residual nit for a future pass: no test drives the OVER-clamp path itself (requested > total); the boundary case pins the bound but not the clamping log/row-count for an over-ask. Sorry for the silent thread -- the fix landed but this reply did not.
